server/middleware/acl.js

```diff
@@ -0,0 +1,556 @@
+        var acl = acls[j];
+        debug('Checking ACL %j against scope %s', acl, s);
+        var aclScopes = acl.scopes || [];
+        if (typeof aclScopes === 'string') {
+          aclScopes = normalizeList(aclScopes);
+        }
+        if (aclScopes.indexOf(s) !== -1) {
+          debug('Matched ACL for scope: %j', s, acl);
+          matchedACLs.push(acl);
+        }
+      }
+    }
+  }
+  return matchedACLs;
+}
+
+/**
+ * Try to find out the protected resource scopes for the given request
+ * @param {Request} req loopback Request
+ * @param {Function} cb Callback function
+ * @returns {*}
+ * @private
+ */
+function identifyScopes(req, scopes, cb) {
+  var routes = normalizeScopeMappings(scopes);
+  var matchedScopes = findMatchedScopes(req, routes);
+  debug('Matched scopes: %j', matchedScopes);
+  return process.nextTick(function() {
+    cb(null, matchedScopes);
+  });
+}
+
+/**
+ * Try to find out the principals for the given request
+ * @param {Request} req HTTP request object
+ * @private
+ */
+function identifyPrincipals(req) {
+  var principals = [{
+    principalType: Principal.ROLE,
+    principalId: '$everyone'
+  }];
+  if (req.accessToken) {
+    if (req.accessToken.userId != null) {
+      principals.push({
+        principalType: Principal.USER,
+        principalId: req.accessToken.userId
+      });
+      principals.push({
+        principalType: Principal.ROLE,
+        principalId: '$authenticated'
+      });
+    }
+    var appId = req.accessToken.appId || req.accessToken.clientId;
+    if (appId != null) {
+      principals.push({
+        principalType: Principal.APP,
+        principalId: appId
+      });
+    }
+  } else {
+    principals.push({
+      principalType: Principal.ROLE,
+      principalId: '$unauthenticated'
+    });
+  }
+  debug('Principals: %j', principals);
+  return principals;
+}
+
+/**
+ * Find matched scopes for the given routes
+ * @param {Request} req HTTP request object
+ * @param {Object[]} routes An array of routes (methods, path)
+ * @returns {Array} Scopes matching the request
+ * @private
+ */
+function findMatchedScopes(req, routes) {
+  var matchedScopes = [];
+  var method = req.method.toLowerCase();
+  var url = req.originalUrl || req.url;
+
+  for (var i = 0, n = routes.length; i < n; i++) {
+    var route = routes[i];
+    debug('Matching %s %s against %j', method, url, route);
+    if (route.methods.indexOf('all') !== -1 ||
+      route.methods.indexOf(method) !== -1) {
+      debug('url: %s, regexp: %s', url, route.regexp);
+      var index = url.indexOf('?');
+      if (index !== -1) {
+        url = url.substring(0, index);
+      }
+      if (route.regexp.test(url)) {
+        matchedScopes.push(route.scope);
+      }
+    }
+  }
+  return matchedScopes;
+}
+
+/**
+ * Calculate the ACL score based on its specifics of principalType and
+ * permission
+ * @param {Object} acl ACL rule
+ * @returns {number}
+ * @private
+ */
+function getACLScore(acl) {
```

Comment:
Does this duplicate the algorithm in loopback core? If so we should have an issue tracking removing that duplication...